### DISCUSSION TITLE
✗ failed to clone repository: unexpected client error: unexpected requesting "https://company@dev.azure.com/company/Client/_git/client/git-upload-pack" status code: 400

### DISCUSSION THREAD

**Discussion by MohamedBenighil** [fluxcd/flux2#3075] [Q&A] [answered]
Hello,

I have a repo having 2 folders :
```
@mohamed.benighil ➜ /workspaces/workdir/Client(GitOps-flux) $ ls
gitops  infrastructure
@mohamed.benighil ➜ /workspaces/workdir/Client (GitOps-flux) $
```

The flux stuff are under `gitops` folder however `infrastructure` folder is for other stuff.

Notes:
1. I am using git scm of azure devops
2. I am using `GitOps-flux` branch (not main). 

My question : 
when i am trying to bootstrap using the following command : 
```
flux bootstrap git \
  --url=https://company@dev.azure.com/company/client/_git/client\
  --username=mohamed.benighil \
  --password=cs****************************************q \
  --token-auth=true \
  --branch=GitOps-flux \
  --path=./gitops/clusters/dev
```

i get the following error : 

```
► cloning branch "GitOps-flux" from Git repository "https://company@dev.azure.com/company/client/_git/client"
✗ failed to clone repository: unexpected client error: unexpected requesting "https://company@dev.azure.com/company/client/_git/client/git-upload-pack" status code: 400
```

I do not know why ? Any help please

**Comment by stefanprodan** [comment, marked as answer]
Flux bootstrap doesn't work with Azure DevOps, we only support providers which have implemented Git protocol v1. Please see the Azure DevOps docs on how to install Flux: https://fluxcd.io/flux/use-cases/azure/#flux-installation-for-azure-devops

**Reply by Tarekdu13** [reply]
Got the same error using gitlab.com.
 failed to push manifests: unexpected client error: unexpected requesting "https://gitlab.com/my-name/repo-name.git/info/refs?service=git-receive-pack" status code: 422

When i used this command:
flux bootstrap gitlab --owner=my-name --repository=repo-name --branch=master --path=clusters/my-cluster --token-auth --personal
I have already exported:
export GITLAB_TOKEN=my-token

Any idea, why ?

**Reply by stefanprodan** [reply]
`status code: 422` means that the request was rejected due to permissions. Does that token have permissions to create repos?

**Reply by Tarekdu13** [reply]
Yes, i have, and i was working on it with flux for like the past 3 hours, everything was working fine until i tried to create a new repo.

**Reply by stefanprodan** [reply]
Then the token can't create repos. What happens if you create the repo from the UI? Just create a blank repo with a README.md in the master branch and rerun bootstrap on it.

**Reply by stefanprodan** [reply]
Or maybe the token has expired, if it stopped working 3 hours ago.

**Reply by Tarekdu13** [reply]
I did as you told me:
1 - Create a repos from UI and bootstrap, seems that it worked just fine.
I tried again after the fail, on a new project and it also worked.
i don't know what to say

**Reply by Tarekdu13** [reply]
i don't think it's a token problem, if flux bootstrap worked after that.

**Reply by repogh** [reply]
Im having the same issue this morning, I have working ADO repo, 3 AKS clusters, all working for last month, then this morning I get:

```
"message": "failed to checkout and determine revision: unable to clone 'https://***@dev.azure.com/***/GitOps-Flux/_git/Fluxv2': unexpected client error: unexpected requesting \"https://***@dev.azure.com/***/GitOps-Flux/_git/Fluxv2/git-upload-pack\" status code: 400",
        "reason": "GitOperationFailed"
  ```      
        
I thought it was PAT token aswell, until I changed them all 3 times, seems something has suddenly broken in backend

Update:

Turns out its a breaking change in new Flux release https://github.com/fluxcd/flux2/releases/tag/v2.0.0-rc.4 which contains breaking changes that could affect 3rd-party controllers that integrate with Flux, which is impacting Azure, changed back to v2.0.0-rc.3 and works fine

**Reply by vliska** [reply]
Any workaround available for rc.4? We also had to rollback (and actually delete gitrepositories and apply them again) to rc.3 (azure devops git repo with https/PAT).

**Reply by stefanprodan** [reply]
A fix for Azure DevOps has been merged and it will be available in RC.5 this week. For now please use RC.3
